internal/controller/reinstall/helper.go

```diff
@@ -324,3 +328,60 @@ func getDataRestorationSummary(
 		fmt.Sprintf("%d resources were successfully restored: %d cluster-identity resources, %d other resources",
 			totalRestoredResources, clusterIDCount, otherCount)), nil
 }
+
+func computeClusterInstanceSpecDiff(clusterInstance *v1alpha1.ClusterInstance) (string, error) {
+
+	lastObserved, ok := clusterInstance.Annotations[v1alpha1.LastClusterInstanceSpecAnnotation]
+	if !ok {
+		return "", nil
+	}
+
+	lastObservedSpec := &v1alpha1.ClusterInstanceSpec{}
+	if err := json.Unmarshal([]byte(lastObserved), lastObservedSpec); err != nil {
+		return "", fmt.Errorf("failed to unmarshal last-observed spec: %w", err)
+	}
+
+	lastObservedSpecJSON, err := json.Marshal(lastObservedSpec)
```

Comment:
Why unmarshal then marshal again? Why not just use `lastObserved`?

Reply:
This is to ensure that we have a valid JSON representation of the `ClusterInstanceSpec` object.